src/catching_show.c

```diff
@@ -0,0 +1,327 @@
+    u16 rnd = LCRandom() % 10;
+    catchingShow->steps = rnd + 5;
+}
+
+static BOOL IsStepCountZero(CatchingShow *catchingShow) {
+    if (--catchingShow->steps == 0) {
+        ResetStepCount(catchingShow);
+        return TRUE;
+    }
+
+    return FALSE;
+}
+
+static enum PalParkEncounterType GetEncounterArea(FieldSystem *fieldSystem, int x, int z) {
+    int behavior = GetMetatileBehavior(fieldSystem, x, z);
+    int quadrant = (x < 32 ? 0 : 1);
+    quadrant += (z < 32 ? 0 : 2);
+    if (MetatileBehavior_IsEncounterGrass(behavior)) {
+        return (enum PalParkEncounterType)(quadrant + PP_ENCTYPE_LAND_MIN);
+    } else if (MetatileBehavior_IsSurfableWater(behavior)) {
+        return (enum PalParkEncounterType)(quadrant + PP_ENCTYPE_WATER_MIN);
+    }
+
+    return PP_ENCTYPE_NONE;
+}
+
+static BOOL TryStartEncounter(FieldSystem *fieldSystem, CatchingShow *catchingShow, int x, int z) {
+    int i;
+    int encounterChance, totalRarity = 0;
+    enum PalParkEncounterType area = GetEncounterArea(fieldSystem, x, z);
+
+    if (area == PP_ENCTYPE_NONE) {
+        return FALSE;
+    }
+
+    // Each uncaught Pokemon has an encounter
+    // chance weight based on its species.
+    // Compute the total weight for the current
+    // area. If it's zero, bail.
+    for (i = 0; i < CATCHING_SHOW_MONS; i++) {
+        if (catchingShow->caughtMonsOrder[i] == 0 && catchingShow->pokemon[i].area == area) {
+            totalRarity += catchingShow->pokemon[i].rarity;
+        }
+    }
+
+    if (totalRarity == 0) {
+        return FALSE;
+    }
+
+    // Prepend an extra bucket of weight 20
+    // for a no-encounter roll.
+    encounterChance = LCRandRange(totalRarity + WEIGHT_NO_ENCOUNTER);
+
+    if (encounterChance < WEIGHT_NO_ENCOUNTER) {
+        return FALSE;
+    }
+
+    encounterChance -= WEIGHT_NO_ENCOUNTER;
+
+    // Using the random number generated,
+    // select the bucket to serve as the
+    // encounter.
+    for (i = 0; i < CATCHING_SHOW_MONS; i++) {
+        if (catchingShow->caughtMonsOrder[i] == 0 && catchingShow->pokemon[i].area == area) {
+            if (encounterChance < catchingShow->pokemon[i].rarity) {
+                catchingShow->currentEncounterIndex = i;
+                return TRUE;
+            } else {
+                encounterChance -= catchingShow->pokemon[i].rarity;
+            }
+        }
+    }
+
+    // In theory, this is unreachable.
+    GF_ASSERT(FALSE);
+    return FALSE;
+}
+
+static void UpdateBattleResultInternal(FieldSystem *fieldSystem, BattleSetup *setup, CatchingShow *catchingShow) {
+    switch (setup->winFlag) {
+    case BATTLE_OUTCOME_MON_CAUGHT:
+        catchingShow->caughtMonsOrder[catchingShow->currentEncounterIndex] = NumMonsCaptured(catchingShow) + 1;
+        break;
+    case BATTLE_OUTCOME_PLAYER_FLED:
+        break;
+    default:
+        GF_ASSERT(FALSE);
+    }
+}
+
+static BattleSetup *SetupEncounter(FieldSystem *fieldSystem, CatchingShow *catchingShow) {
```

Comment:
Naming:
BattleSetup_NewInit_PalPark (or something similar, since BattleSetup is todo to be renamed to BattleDataTransferObject)